From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Wed, 22 Apr 2026 23:48:33 +0800
Subject: [PATCH] Leaf Optimize SetLookAndInteract and
 NearestVisibleLivingEntities

Co-authored by: Taiyou06 <kaandindar21@gmail.com>
As part of: Leaf (https://github.com/Winds-Studio/Leaf/blob/86c87e1ba03840470ed21d6dc872463a9d579edf/leaf-server/minecraft-patches/features/0212-Optimize-SetLookAndInteract-and-NearestVisibleLiving.patch)
Licensed under: MIT (https://github.com/Winds-Studio/Leaf/blob/86c87e1ba03840470ed21d6dc872463a9d579edf/licenses/MIT.txt)

diff --git a/net/minecraft/world/entity/ai/behavior/SetLookAndInteract.java b/net/minecraft/world/entity/ai/behavior/SetLookAndInteract.java
index aeeb3db038a2b861d1dc667379fd971b28476bad..ca6dbfc7b29547effc79e788328d75c549071d60 100644
--- a/net/minecraft/world/entity/ai/behavior/SetLookAndInteract.java
+++ b/net/minecraft/world/entity/ai/behavior/SetLookAndInteract.java
@@ -20,7 +20,7 @@ public class SetLookAndInteract {
                     i,
                     (lookTarget, interactionTarget, nearestEntities) -> (level, body, timestamp) -> {
                         Optional<LivingEntity> closest = i.<NearestVisibleLivingEntities>get(nearestEntities)
-                            .findClosest(e -> e.distanceToSqr(body) <= interactionRangeSqr && e.is(type));
+                            .findClosest(e -> e.is(type) && e.distanceToSqr(body) <= interactionRangeSqr); // Leaf - Optimize SetLookAndInteract and NearestVisibleLivingEntities
                         if (closest.isEmpty()) {
                             return false;
                         } else {
diff --git a/net/minecraft/world/entity/ai/memory/NearestVisibleLivingEntities.java b/net/minecraft/world/entity/ai/memory/NearestVisibleLivingEntities.java
index 0d706560d605299de6d4e9fe0554c6210715a7aa..aa50edb208e2b690e960da9d10856ba953c420e3 100644
--- a/net/minecraft/world/entity/ai/memory/NearestVisibleLivingEntities.java
+++ b/net/minecraft/world/entity/ai/memory/NearestVisibleLivingEntities.java
@@ -38,11 +38,18 @@ public class NearestVisibleLivingEntities {
     }
 
     public Optional<LivingEntity> findClosest(final Predicate<LivingEntity> filter) {
-        for (LivingEntity nearbyEntity : this.nearbyEntities) {
+        // Leaf start - Optimize SetLookAndInteract and NearestVisibleLivingEntities
+        if (this.nearbyEntities.isEmpty()) {
+            return Optional.empty();
+        }
+
+        for (int i = 0, size = this.nearbyEntities.size(); i < size; i++) {
+            LivingEntity nearbyEntity = this.nearbyEntities.get(i);
             if (filter.test(nearbyEntity) && this.lineOfSightTest.test(nearbyEntity)) {
                 return Optional.of(nearbyEntity);
             }
         }
+        // Leaf end - Optimize SetLookAndInteract and NearestVisibleLivingEntities
 
         return Optional.empty();
     }
@@ -55,6 +62,19 @@ public class NearestVisibleLivingEntities {
         return this.nearbyEntities.stream().filter(entity -> filter.test(entity) && this.lineOfSightTest.test(entity));
     }
 
+    // Leaf start - Optimize SetLookAndInteract and NearestVisibleLivingEntities
+    public List<LivingEntity> findWithList(Predicate<LivingEntity> predicate) {
+        it.unimi.dsi.fastutil.objects.ObjectList<LivingEntity> result = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
+        for (int i = 0, size = this.nearbyEntities.size(); i < size; i++) {
+            LivingEntity entity = this.nearbyEntities.get(i);
+            if (predicate.test(entity) && this.lineOfSightTest.test(entity)) {
+                result.add(entity);
+            }
+        }
+        return result;
+    }
+    // Leaf end - Optimize SetLookAndInteract and NearestVisibleLivingEntities
+
     public boolean contains(final LivingEntity targetEntity) {
         return this.nearbyEntities.contains(targetEntity) && this.lineOfSightTest.test(targetEntity);
     }
